File: java/src/test/java/challenges/aoc2024/day2/function/ReactorSafetyPredicateTest.java
package challenges.aoc2024.day2.function;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import challenges.aoc2024.day2.model.ReactorReport;

class ReactorSafetyPredicateTest {

    @ParameterizedTest
    @CsvSource({
        "7 6 4 2 1, true",
        "1 2 7 8 9, false",
        "9 7 6 2 1, false",
        "1 3 2 4 5, false",
        "8 6 4 4 1, false",
        "1 3 6 7 9, true"
    })
    void shouldDetermineIfReactorIsSafe(String inputReport, boolean expectedSafetyOutcome) {
        // Given
        final ReactorReport reactorReport = ReactorReport.of(inputReport);
        final ReactorSafetyPredicate predicate = new ReactorSafetyPredicate();

        // When
        final boolean isSafe = predicate.test(reactorReport);

        // Then
        assertEquals(expectedSafetyOutcome, isSafe);
    }
}

File: scala/src/test/scala/challenges/aoc2024/day2/function/ReactorSafetyPredicateTest.scala
package challenges.aoc2024.day2.function

import challenges.aoc2024.day2.model.ReactorReport
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.prop.TableDrivenPropertyChecks.forAll
import org.scalatest.prop.TableFor2
import org.scalatest.prop.Tables.Table

class ReactorSafetyPredicateTest extends AnyFunSuite {

  val noDampenerCases: TableFor2[String, Boolean] = Table(
    ("inputReport", "expectedSafetyOutcome"),
    ("7 6 4 2 1", true),
    ("1 2 7 8 9", false),
    ("9 7 6 2 1", false),
    ("1 3 2 4 5", false),
    ("8 6 4 4 1", false),
    ("1 3 6 7 9", true)
  )
  test("should determine if reactor is safe with no dampener active"):
    forAll(noDampenerCases) { (inputReport: String, expectedSafetyOutcome: Boolean) =>
      // Given
      val reactorReport: ReactorReport = ReactorReport.of(inputReport)
      val predicate: ReactorSafetyPredicate = new ReactorSafetyPredicate(false)

      // When
      val isSafe: Boolean = predicate.apply(reactorReport)

      // Then
      assert(expectedSafetyOutcome == isSafe)
    }

  val dampenerCases: TableFor2[String, Boolean] = Table(
    ("inputReport", "expectedSafetyOutcome"),
    ("7 6 4 2 1", true),
    ("1 2 7 8 9", false),
    ("9 7 6 2 1", false),
    ("1 3 2 4 5", true),
    ("8 6 4 4 1", true),
    ("1 3 6 7 9", true)
  )
  test("should determine if reactor is safe with dampener active"):
    forAll(dampenerCases) { (inputReport: String, expectedSafetyOutcome: Boolean) =>
      // Given
      val reactorReport: ReactorReport = ReactorReport.of(inputReport)
      val predicate: ReactorSafetyPredicate = new ReactorSafetyPredicate(true)

      // When
      val isSafe: Boolean = predicate.apply(reactorReport)

      // Then
      assert(expectedSafetyOutcome == isSafe)
    }
}
